streams/src/test/java/org/apache/kafka/streams/state/internals/CachingKeyValueStoreTest.java

```diff
@@ -107,32 +108,98 @@ public void shouldSetFlushListener() {
     public void shouldAvoidFlushingDeletionsWithoutDirtyKeys() {
         final int added = addItemsToCache();
         // all dirty entries should have been flushed
-        assertEquals(added, underlyingStore.approximateNumEntries());
+        assertEquals(added, keyValueStore.approximateNumEntries());
         assertEquals(added, cacheFlushListener.forwarded.size());
 
         store.put(bytesKey("key"), bytesValue("value"));
-        assertEquals(added, underlyingStore.approximateNumEntries());
+        assertEquals(added, keyValueStore.approximateNumEntries());
         assertEquals(added, cacheFlushListener.forwarded.size());
 
         store.put(bytesKey("key"), null);
         store.flush();
-        assertEquals(added, underlyingStore.approximateNumEntries());
+        assertEquals(added, keyValueStore.approximateNumEntries());
         assertEquals(added, cacheFlushListener.forwarded.size());
     }
 
+    @Test
+    public void shouldCloseWrappedStoreAfterErrorDuringCacheFlush() {
+        setUpCloseTests();
+        cache.flush(CACHE_NAMESPACE);
+        EasyMock.expectLastCall().andThrow(new NullPointerException("Simulating an error on flush"));
+        EasyMock.replay(cache);
+        EasyMock.reset(keyValueStore);
+        keyValueStore.close();
+        EasyMock.replay(keyValueStore);
+
+        try {
+            store.close();
+        } catch (final NullPointerException npe) {
+            EasyMock.verify(keyValueStore);
```

Comment:
Should we verify regardless of the NPE?

Reply:
I changed `NullPointerException` to `RuntimeException` (also due to the changes to not throw in a finally block). The test verifies that the underlying state store is closed if a runtime exception is thrown. I think that it is fine now.

Reply:
Ok, I was just concerned that if the exception weren't thrown, we would be verifying nothing, and the test would pass silently, but if you prefer this structure, then I'm fine with it.

Reply:
The missing piece is a unit test that verifies that `close()` of the underlying store is called when `close()` of the caching store is called.

Reply:
Wait, there are tests that do that.